tfhe/src/core_crypto/fft_impl/common.rs

```diff
@@ -1,44 +1,30 @@
 use crate::core_crypto::commons::math::torus::UnsignedTorus;
 use crate::core_crypto::commons::numeric::{CastInto, UnsignedInteger};
 use crate::core_crypto::commons::parameters::{
-    DecompositionBaseLog, DecompositionLevelCount, GlweSize, LutCountLog, LweDimension,
-    ModulusSwitchOffset, PolynomialSize,
+    DecompositionBaseLog, DecompositionLevelCount, GlweSize, LweDimension, PolynomialSize,
 };
 use crate::core_crypto::commons::traits::Container;
 use crate::core_crypto::entities::*;
-use crate::core_crypto::prelude::ContainerMut;
+use crate::core_crypto::prelude::{CiphertextModulusLog, ContainerMut};
 use dyn_stack::{PodStack, SizeOverflow, StackReq};
 
-/// This function switches modulus for a single coefficient of a ciphertext,
-/// only in the context of a PBS
-///
-/// - offset: the number of msb discarded
-/// - lut_count_log: the right padding
-///
-/// # Note
-///
-/// If you are switching to a modulus of $2N$ then this function may return the value $2N$ while a
-/// "true" modulus switch would return $0$ in that case. It turns out that this is not affecting
-/// other parts of the code relying on the modulus switch (as a rotation by $2N$ is effectively the
-/// same as rotation by $0$ for polynomials of size $N$ in the ring $X^N+1$) but it could be
-/// problematic for code requiring an output in the expected $[0; 2N[$ range. Also this saves a few
-/// instructions which can add up when this is being called hundreds or thousands of times per PBS.
+/// This function switches modulus for a single coefficient of a ciphertext
 pub fn fast_pbs_modulus_switch<Scalar: UnsignedTorus + CastInto<usize>>(
     input: Scalar,
-    poly_size: PolynomialSize,
-    offset: ModulusSwitchOffset,
-    lut_count_log: LutCountLog,
+    log_modulus: CiphertextModulusLog,
 ) -> usize {
     // First, do the left shift (we discard the offset msb)
-    let mut output = input << offset.0;
+    let mut output = input;
+
     // Start doing the right shift
-    output >>= Scalar::BITS - poly_size.log2().0 - 2 + lut_count_log.0;
+    output >>= Scalar::BITS - log_modulus.0 - 1;
     // Do the rounding
     output += Scalar::ONE;
-    // Finish the right shift
-    output >>= 1;
-    // Apply the lsb padding
-    output <<= lut_count_log.0;
+
+    output <<= Scalar::BITS - log_modulus.0 - 1;
+
+    output >>= Scalar::BITS - log_modulus.0;
+
     <Scalar as CastInto<usize>>::cast_into(output)
```

Comment:
how about shifting 1 to the right position, adding it, and shifting right once to discard anything below a certain bit ?

Reply:
don't hesitate to code golf a bit on godbolt with -O3

Reply:
The code is simpler the way you described so I changed it
It's the same number of instructions on godbolt
Anyway I don't think saving a few instructions here will visibly affect PBS performance

Reply:
for sure, it's just one of the rare bit tricks functions where we can actually easily reason about that 😅